### ISSUE TITLE
Memory Problems During Inference with GNPE

### ISSUE THREAD

**Issue by nihargupte-ph** [dingo-gw/dingo#83]
For some reason when calling the inference code, CUDA runs out of memory. 

Eg
```
dingo_posterier = dingo.gw.inference.sampling_functions.sample_posterier_of_injection(strain_data, model=main_pm, model_init=time_pm, device='cuda', num_samples=50_000, batch_size=10_000)
```
Will run out of memory on a 40GB A100 TPU. 

Temporary workaround: 

```
def sample_posterier(strain_data, main_pm, time_pm):
    def memory_wrapper(strain_data, main_pm, time_pm):
        dingo_posterier = dingo.gw.inference.sampling_functions.sample_posterier_of_injection(strain_data, model=main_pm, model_init=time_pm, device='cuda', num_samples=10_000)
        return dingo_posterier

    # 50,000 samples samples 5 times from the network, should be deterministic ? Done for memory reasons
    for i in range(5):
        dingo_posterier_temp = memory_wrapper(strain_data, main_pm, time_pm)
        if not 'dingo_posterier' in locals():
            dingo_posterier = {k:v.cpu().numpy() for k,v in dingo_posterier_temp.items()}
        else:
            dingo_posterier = {k:np.concatenate((dingo_posterier[k], dingo_posterier_temp[k].cpu().numpy()), axis=None) for k,v in dingo_posterier.items()}

    return dingo_posterier

dingo_posterier = sample_posterier(strain_data, main_pm, time_pm)
```
This samples 50,000 points from the posterior by sampling 10,000 points 5 times and concatenating. It does not run out of CUDA memory. I found this odd at first since I'd expect this is exactly what batching should do but I see why now it is slightly different (see below). 

I debugged the program profilling it with a realtime GPU memory reporter here's what I found:

In the ResetSample transform in the inference_transforms.py the waveform gets cloned with `sample["waveform"] = sample["waveform_"].clone()` to make sure the WF doesn't get edited through the transforms. I think later down the line this may get deleted but freeing up the memory may be useful for the rest of the transform pipeline.

In the `gnpe_pre_transforms` eventually FrequencyDomain.time_translate_data gets called. Here large tensors like: `cos_txf = torch.empty((batch_size, Nd, Nf), device=data.device)` are created. I found that this batch_size is not the 10_000 passed to the inference api function but rather the 50_000 of the number of samples requested. This is effectively what is causing the memory shutdown. It also explains why my memory wrapper works when using a size of 10_000 but a batch_size of 10_000 does not. I guess this makes sense since the batching only applies to sampling from the main network posterior and not the time network posterior. So it is passing 50_000 samples as context to the main network and causing a memory shutdown. Could also use pytorch dataparallel but I think that is not a great solution here.

I think the reason I am the first to run into this is my tensor sizes are torch.Size([50000, 2, 3, 16225]) the 16225 is double 8193 since I am looking at frequencies from 20 to 2048 instead of 20 to 1024.

TLDR: I don't think it's actually a memory leak persay but sampling a large number of points does run into memory issues. Batching the FrequencyDomain.time_translate_data function may work but it may be tricky so I figure just using the workaround should be ok for now.
